osutil/group.go

```diff
@@ -20,35 +20,112 @@
 package osutil
 
 import (
+	"bytes"
+	"fmt"
+	"os/exec"
 	"os/user"
 	"strconv"
 )
 
-// TODO: the builtin os/user functions only look at /etc/passwd and /etc/group
-// which is fine for our purposes today. In the future we may want to support
-// lookups in extrausers, which is configured via nsswitch.conf. Since snapd
-// does not support being built with cgo itself, when we want to support
-// extrausers here, we can convert these to do the equivalent of:
-//
-//   getent passwd <user> | cut -d : -f 3
-//   getent group <group> | cut -d : -f 3
-
-// FindUid returns the identifier of the given UNIX user name.
+// FindUid returns the identifier of the given UNIX user name with no getent
+// fallback
 func FindUid(username string) (uint64, error) {
-	user, err := user.Lookup(username)
+	return findUid(username)
+}
+
+// FindGid returns the identifier of the given UNIX group name with no getent
+// fallback
+func FindGid(groupname string) (uint64, error) {
+	return findGid(groupname)
+}
+
+// getent returns the identifier of the given UNIX user or group name as
+// determined by the specified database
+func getent(name string, database string) (uint64, error) {
+	if database != "passwd" && database != "group" {
+		return 0, fmt.Errorf(`unsupported getent database "%q"`, database)
+	}
+
+	cmdStr := []string{
+		"getent",
+		database,
+		name,
+	}
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		// according to getent(1) the exit value of "2" means:
+		//    One or more supplied key could not be found in
+		//    the database.
+		exitCode, _ := ExitCode(err)
```

Comment:
My understanding was that this is Linux-specific (which is why I didn't do it to begin with) but clearly you are ok with this since it is already in osutil/exitcode.go. FYI, https://golang.org/pkg/os/#ProcessState.ExitCode which is in golang 1.12 AIUI (https://github.com/golang/go/issues/26539).